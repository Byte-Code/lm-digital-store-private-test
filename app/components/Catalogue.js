import React, { Component, PropTypes } from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Map, Range } from 'immutable';
import styled from 'styled-components';

import ProductBadge from './ProductBadge';
import ProductSlider from './ProductSlider';

const Header = styled.div`
  width: 100%;
  height: 230px;
  background: #f7f7f7;
  display: flex;
  justify-content: center;
  align-items: center;
  &>h1 {
    font-size: 48px;
  }
`;

const Slide = styled.div`
  margin-right: 20px;
  width: 405px;
  flex-direction: column;
  &>div:first-child {
    margin-bottom: 60px;
  }
`;

const Filters = styled.div`
  height: 75px;
  background: #efefef;
  margin-bottom: 115px;
`;

export default class Catalogue extends Component {
  static propTypes = {
    params: PropTypes.shape({ categoryCode: PropTypes.string.isRequired }).isRequired,
    requestFetchCategory: PropTypes.func.isRequired,
    categoryInfo: ImmutablePropTypes.map
  }

  static defaultProps = {
    categoryInfo: Map()
  }

  componentDidMount() {
    const {
      params: { categoryCode },
      requestFetchCategory
    } = this.props;
    requestFetchCategory(categoryCode);
  }

  chunkItemList(chunkSize) {
    const itemList = this.props.categoryInfo.get('itemList');
    return Range(0, itemList.count(), chunkSize)
      .map(chunkStart => itemList.slice(chunkStart, chunkStart + chunkSize));
  }

  render() {
    const { categoryInfo } = this.props;

    if (categoryInfo.isEmpty()) {
      return null;
    }

    const catName = categoryInfo.get('name');
    const itemList = this.chunkItemList(2);
    const sliderItems = itemList.map(item => {
      const key = `${item.getIn([0, 'code'])}-${item.getIn([1, 'code'])}`;
      return (
        <Slide key={key}>
          <ProductBadge productInfo={item.get(0)} />
          <ProductBadge productInfo={item.get(1)} />
        </Slide>
      );
    }
  );

    return (
      <div>
        <Header>
          <h1>{catName}</h1>
        </Header>
        <Filters>Filters here</Filters>
        <ProductSlider>
          {sliderItems}
        </ProductSlider>
      </div>
    );
  }
}
