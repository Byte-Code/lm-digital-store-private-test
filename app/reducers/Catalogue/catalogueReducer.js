import { OrderedSet } from 'immutable';

export default function catalogueReducer(state = OrderedSet(), action) {
  const { result } = action;
  switch (action.type) {
    case 'SUCCESS_FETCH_PRODUCTLIST':
      return result;
    case 'CLEAR_PRODUCT_LIST':
      return OrderedSet();
    default:
      return state;
  }
}
