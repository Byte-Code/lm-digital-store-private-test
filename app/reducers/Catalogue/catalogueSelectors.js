/* eslint-disable */
export function getProductList(state) {
  return state;
}
/* eslint-disable */
