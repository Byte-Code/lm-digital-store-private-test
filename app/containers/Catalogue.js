import { connect } from 'react-redux';

import Catalogue from '../components/Catalogue';
import { requestFetchCategory } from '../actions/categoryActions';
import { getCategory } from '../reducers/selectors';

const mapStateToProps = (state, ownProps) => ({
  categoryInfo: getCategory(state, ownProps.params.categoryCode)
});

const mapDispatchToProps = {
  requestFetchCategory
};

export default connect(mapStateToProps, mapDispatchToProps)(Catalogue);
