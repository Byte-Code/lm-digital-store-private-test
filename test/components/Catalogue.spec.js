import React from 'react';
import { fromJS, Map, List } from 'immutable';
import { shallow, mount } from 'enzyme';

import Catalogue from '../../app/components/Catalogue';

const params = { categoryCode: 'CAT123' };
const requestFetchCategory = jest.fn();
const initFilters = jest.fn();
const products = fromJS([
  { code: 1, mainImage: 'image1', name: 'product1' },
  { code: 2, mainImage: 'image2', name: 'product2' },
  { code: 3, mainImage: 'image3', name: 'product3' }
]);
const categoryInfo = fromJS({
  name: 'cat123',
  sellingAidsProducts: List(),
  facetFilters: List()
});
const router = { location: {} };
const filterMap = Map();

describe('catalogue', () => {
  it('should render null when categoryInfo is empty or undefined', () => {
    const result = shallow(
      <Catalogue
        params={params}
        filterMap={filterMap}
        products={products}
        requestFetchCategory={requestFetchCategory}
        router={router}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        initFilters={() => {}}
        applyFilters={() => {}}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should render properly when categoryInfo is not empty', () => {
    const result = shallow(
      <Catalogue
        params={params}
        categoryInfo={categoryInfo}
        filterMap={filterMap}
        products={products}
        requestFetchCategory={requestFetchCategory}
        router={router}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        initFilters={() => {}}
        applyFilters={() => {}}
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should call requestFetchCategory after mount', () => {
    mount(
      <Catalogue
        params={params}
        categoryInfo={categoryInfo}
        filterMap={filterMap}
        products={products}
        requestFetchCategory={requestFetchCategory}
        router={router}
        initFilters={initFilters}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        applyFilters={() => {}}
      />
    );
    expect(initFilters).toHaveBeenCalled();
    expect(requestFetchCategory).toHaveBeenCalledWith('CAT123');
  });

  it('should call initFilters and requestFetchCategory when categoryCode changes', () => {
    const result = mount(
      <Catalogue
        params={params}
        categoryInfo={categoryInfo}
        filterMap={filterMap}
        products={products}
        router={router}
        requestFetchCategory={requestFetchCategory}
        initFilters={initFilters}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        applyFilters={() => {}}
      />
    );
    const nextProps = { params: { categoryCode: 'CAT456' } };
    result.setProps(nextProps);
    expect(initFilters).toHaveBeenCalled();
    expect(requestFetchCategory).toHaveBeenCalledWith('CAT456');
  });
});
