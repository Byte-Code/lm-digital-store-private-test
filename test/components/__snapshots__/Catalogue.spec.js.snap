// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalogue should render null when categoryInfo is empty or undefined 1`] = `null`;

exports[`catalogue should render properly when categoryInfo is not empty 1`] = `
<div>
  <glamorous(div)>
    <h1>
      cat123
    </h1>
  </glamorous(div)>
  <SellingAidsBadge
    activeAid=""
    onToggle={[Function]}
    sellingAids={Object {}}
  />
  <FilterBar
    filterGroups={Array []}
    filterMap={Object {}}
    resetFilters={[Function]}
    toggleAvailability={[Function]}
    toggleFilter={[Function]}
  />
  <glamorous(div)>
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="product/1"
    >
      <ProductBadge
        handleClick={[Function]}
        productInfo={
          Object {
            "code": 1,
            "mainImage": "image1",
            "name": "product1",
          }
        }
      />
    </Link>
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="product/2"
    >
      <ProductBadge
        handleClick={[Function]}
        productInfo={
          Object {
            "code": 2,
            "mainImage": "image2",
            "name": "product2",
          }
        }
      />
    </Link>
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="product/3"
    >
      <ProductBadge
        handleClick={[Function]}
        productInfo={
          Object {
            "code": 3,
            "mainImage": "image3",
            "name": "product3",
          }
        }
      />
    </Link>
  </glamorous(div)>
</div>
`;
